plugins/devlink/fu-self-test.c

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright 2025 NVIDIA Corporation & Affiliates
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-context-private.h"
+#include "fu-devlink-component.h"
+#include "fu-devlink-device.h"
+
+typedef struct {
+	guint device_id;
+} FuDevlinkNetdevsim;
+
+static gboolean
+fu_devlink_file_write_helper(const gchar *path, const guint value, GError **error)
+{
+	g_autofree gchar *value_str = g_strdup_printf("%u", value);
+	g_autoptr(FuIOChannel) io = NULL;
+
+	/* check if file exists first */
+	if (!g_file_test(path, G_FILE_TEST_IS_REGULAR)) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND, "file not found: %s", path);
+		return FALSE;
+	}
+
+	io = fu_io_channel_new_file(path, FU_IO_CHANNEL_OPEN_FLAG_WRITE, error);
+	if (io == NULL)
+		return FALSE;
+
+	return fu_io_channel_write_raw(io,
+				       (const guint8 *)value_str,
+				       strlen(value_str),
+				       1000,
+				       FU_IO_CHANNEL_FLAG_NONE,
+				       error);
+}
+
+static gboolean
+fu_devlink_load_kernel_module_helper(const gchar *module_name, GError **error)
+{
+	gint exit_status = 0;
+	g_autofree gchar *stderr_output = NULL;
+	const gchar *argv[] = {"modprobe", module_name, NULL};
```

Comment:
will this work without being root? I think a better thing would be a g_skip_test() with what the user needs to do.

Reply:
for modprobe you have to be root. But you have to be root when you create netdevsim. But for both, eventually g_skip_test is done:
`ok 1 /devlink/netdevsim/device # SKIP Failed to create netdevsim device: failed to open /sys/bus/netdevsim/new_device: Permission denied`
So it is how you want it, correct?

Reply:
I don;t think we should be modprobing inside the self tests at all; it's the sort of thing that's going to light up a security scanner like christmas day. I think the right thing to do is to exit the test and just show the `failed to load netdevsim module` message.

Reply:
> I also think we need to modprode that module in at least one (or more?) of the CI targets, at the moment the CI coverage isn't quite what I wanted to see: https://app.codecov.io/gh/fwupd/fwupd/pull/9053

How can I do that? If I can't could you please? Will try to add more paths to the selftest to make the coverage a bit better.

Reply:
> I think it might be as easy as adding the modprobe line in contrib/ci/arch.sh (or whatever distro you'd like) -- and then that would increase the coverage score dramatically automatically.

That did not work: "modprobe: FATAL: Module netdevsim not found in directory /lib/modules/6.11.0-1018-azure". I put it to Fedora. Looks like the running kernel is not Fedora. I guess you do not run per-distro kernels in CI, do you.

Reply:
We're running all the CI in containers not bare metal.  So we don't get to control the kernel modules loaded AFAIK.